In [25]:
import sklearn
from sklearn.datasets import load_iris

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
['setosa' 'versicolor' 'virginica']
<class 'numpy.ndarray'>
(150, 4)
(150,)
KNeighborsClassifier(n_neighbors=1)


array([2, 1])

#saves iris as a "bunch" object in sklearn - special type of array?
#the .data appendix is an "attribute" of the bunch

In [ ]:
iris = load_iris()

type(iris)

#print(iris.data)

#the feature names are the column names

In [ ]:
print(iris.feature_names)

print(iris.target)

#the .target is the feature to be predicted (the response)

In [ ]:
print(iris.target_names)

#the .target_names are the names of the responses to be predicted (the species)
#features and responses are separate objects (ie, iris.data vs iris.target)
#features should be numeric and take the form of a numpy array

In [ ]:
print(type(iris.data))

#features and responses should have specific shapes

In [ ]:
x = iris.data
y = iris.target

#storing the feature array and response array in variables x and y resp.

In [ ]:
print(iris.data.shape)
print(iris.target.shape)

#checks the shape of the array; the data array is RxC, with columns being features
#and the target array is a vector with one feature: the response

#knn (k nearest neighbors) classification modeling

#pick a value for k, search for the k observations in the training data that
#are nearest to the dimensions of the test row (unknown iris)
#use the most popular response value for the k nearest neighbors (mode of the
#responses for each observation) as the predicted response for the unknown
#sample. this is basically a location function.

#knn model can generate classification maps (think, predicted responses around
#clusters of similar observations, and the area that the response space
#encompasses; the map changes as k changes; the boundaries between the spaces are known
#as response boundaries


#import the class to use

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

#instantiate the estimator: make an instance of the model

In [ ]:
knn = KNeighborsClassifier(n_neighbors=1)

#passes the argument number of neighbors = 1 to the kneighborsclassifier function;
#this function creates the object knn
#other parameters can be passed into the function as well

In [ ]:
print(knn)

knn.fit(x, y)

#fits the model to the data (training)

#knn.predict([3, 5, 4, 2])

#this predicts the response (classification/species) for the sample observation
#based on the object (model) knn (which models based on k=1)

In [ ]:
x_new = [[3, 5, 4, 2], [5, 4, 3, 2]]
knn.predict(x_new)
#% md
#defines a pair of observations in x_new and predicts their placements
#the first obs returns 2, and the second returns 1

In [ ]:
knn_5 = KNeighborsClassifier(n_neighbors=5)

knn_5.fit(x,y)
knn_5.predict(x_new)

#defines a new model, knn_5, with parameter n_neighbors=5; this new "tuned"
#model returns different results from knn with k = 1

In [26]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()
logreg.fit(x, y)
logreg.predict(x_new)

#logreg model for comparison with knn

c:\users\joe bilious\appdata\local\programs\python\python38\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


array([0, 0])

#model evaluation procedures to determine which model and which parameters
#to use

#1 train and test on the entire dataset
#basically just check predicted response values against true response values;
#modeled values vs real. duh.

In [ ]:
logreg.predict(x)
y_pred = logreg.predict(x)
len(y_pred)

#instantiates the predict function with x (iris.data, the observations)
#and assigns it to y_pred. then prints the number of observations (150)

In [60]:
print(y_pred)

from sklearn import metrics

print(metrics.accuracy_score(y, y_pred))

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 1 1 1
 1 1 1 2 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 1 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
0.9733333333333334


#imports the metrics module; the accuracy_score function compares y (iris.target)
#with the model generated by logreg.predict(x) (which returns response values
#in a vector - the estimated feature of each observation); compares y (real)
#with x (estimated)

In [59]:
y_pred2 = KNeighborsClassifier(n_neighbors=4)
y_pred2.fit(x,y)
print(y_pred2)

#print(metrics.accuracy_score(y, y_pred2))
metrics.accuracy_score(y, x)

KNeighborsClassifier(n_neighbors=4)


ValueError: Classification metrics can't handle a mix of multiclass and continuous-multioutput targets

#this is supposed to return an accuracy score for the knn model, but
#it's not working right now for some reason